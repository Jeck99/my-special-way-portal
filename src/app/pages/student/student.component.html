<div class="main-content">
  <div class="container-fluid">
    <div class="header-menu-container msw-table-tools-header">
      <div class="example-header">
        <mat-form-field floatLabel="never" color="accent">
          <input matInput #filter placeholder="Filter issues">
        </mat-form-field>
      </div>
      <div class="example-header msw-table-tools-header">
        <button mat-raised-button color="primary" routerLink="/student/new">
          <i class="material-icons">
            add
          </i>
          הוספת תלמיד חדש
        </button>
      </div>
    </div>
    <div class="row">
      <div class="col-lg-12 col-md-12 col-sm-12">
        <!--matSortActive="userName" matSortDisableClear matSortDirection="asc"-->
        <mat-table [dataSource]="dataSource" class="msw-table" matSort matSortDisableClear>
          <!-- Number Column -->
          <ng-container matColumnDef="studentName">
            <mat-header-cell *matHeaderCellDef>שם</mat-header-cell>
            <mat-cell *matCellDef="let row">{{ row.firstname }} {{ row.lastname }}</mat-cell>
          </ng-container>

          <!-- Title Column -->
          <ng-container matColumnDef="gradeId">
            <mat-header-cell *matHeaderCellDef mat-sort-header>כיתה</mat-header-cell>
            <mat-cell *matCellDef="let row">{{(row.Class && row.Class.name) || '' }}</mat-cell>
          </ng-container>

          <!-- username Column -->
          <ng-container matColumnDef="username">
            <mat-header-cell *matHeaderCellDef mat-sort-header>שם משתמש</mat-header-cell>
            <mat-cell *matCellDef="let row">{{ row.username }}</mat-cell>
          </ng-container>

          <!-- State Column -->
          <ng-container matColumnDef="personalArea">
            <mat-header-cell *matHeaderCellDef>אזור אישי</mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button mat-icon-button color="primary" [routerLink]="['/student', row._id]">כניסה</button>
            </mat-cell>
          </ng-container>

          <!--&lt;!&ndash; State Column &ndash;&gt;-->
          <ng-container matColumnDef="editDetails">
            <mat-header-cell *matHeaderCellDef>עריכת פרטים</mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button mat-icon-button color="primary" [routerLink]="['/student', row._id]">ערוך</button>
            </mat-cell>
          </ng-container>

          <!--&lt;!&ndash; State Column &ndash;&gt;-->
          <ng-container matColumnDef="deleteUser">-->
            <mat-header-cell *matHeaderCellDef>מחיקת משתמש</mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button mat-icon-button color="primary" (click)="deleteStudent(row._id, row.firstname, row.lastname, row.gradeId)"><i class="material-icons"> delete </i></button>
            </mat-cell>
          </ng-container>

          <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
          <mat-row *matRowDef="let row; columns: displayedColumns;"></mat-row>
        </mat-table>

        <mat-paginator #paginator [length]="1" [pageIndex]="0" [pageSize]="10" [pageSizeOptions]="[5, 10, 25, 100]">
        </mat-paginator>
      </div>
    </div>
  </div>
</div>
