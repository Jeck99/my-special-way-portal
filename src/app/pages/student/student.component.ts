import {Component, OnInit, AfterViewInit, ViewChild, ElementRef} from '@angular/core';
import { MatTableDataSource, MatPaginator, MatSort, MatDialog } from '@angular/material';
import { DeleteStudentDialogComponent } from './dialogs/delete/delete-student.dialog';
import { merge } from 'rxjs/observable/merge';
import { of as observableOf } from 'rxjs/observable/of';
import { catchError } from 'rxjs/operators/catchError';
import { map } from 'rxjs/operators/map';
import { startWith } from 'rxjs/operators/startWith';
import { switchMap } from 'rxjs/operators/switchMap';
import * as _ from 'lodash';
import { StudentService } from './services/student.graphql.service';
import { AddStudentDialogComponent } from './dialogs/add/add-student.dialog';
import Student from '../../models/student.model';
import { UpdateStudentDialogComponent } from './dialogs/update/update-student.dialog';

@Component({
  selector: 'app-student',
  templateUrl: './student.component.html',
  styleUrls: ['./student.component.scss'],
})
export class StudentComponent implements OnInit, AfterViewInit {

  displayedColumns = ['studentName', 'gradeId', 'userName', 'personalArea', 'editDetails', 'deleteUser'];
  dataSource = new MatTableDataSource();
  resultsLength = 0;

  @ViewChild(MatPaginator) paginator: MatPaginator;
  @ViewChild(MatSort) sort: MatSort;
  @ViewChild('table') table: ElementRef;

  constructor(
    private studentService: StudentService,
    public dialog: MatDialog,
  ) {}

  ngOnInit(): void {
    this.sort.sortChange.subscribe(() => this.paginator.pageIndex = 0);
    merge(this.sort.sortChange, this.paginator.page)
      .pipe(
        startWith({}),
        switchMap(() => {
          return this.studentService.getAllStudents();
        }),
        map((data) => {
          return data.data.students;
        }),
        catchError((err) => {
          console.log(err);
          return observableOf([]);
        }),
    ).subscribe((data) => this.dataSource.data = [...data]);

  }

  ngAfterViewInit() {
    this.dataSource.paginator = this.paginator;
    this.dataSource.sort = this.sort;
  }

  addNewStudent() {
    const dialogRef = this.dialog.open(AddStudentDialogComponent, {
      data: { student: Student },
      height: '600px',
      width: '350px',
    });
    dialogRef.afterClosed().subscribe((data) => {
      if (data) {
        const newStudent: Student = this._createNewStudent(data);
        this.studentService.create(newStudent)
          .then(() => {
            this.dataSource.data.push(newStudent);
            this.dataSource.paginator = this.paginator;
          });
      }
    });
  }

  deleteStudent(id: number, firstName: string, lastName: string, gradeId: string) {
    const dialogRef = this.dialog.open(DeleteStudentDialogComponent, {
      data: {id, firstName, lastName, gradeId},
    });

    dialogRef.afterClosed().subscribe((result) => {
      if (result === true) {
        this.studentService.delete(id)
        .then(() => {
          const index = _.findIndex(this.dataSource.data, (user) => user.id === id );
          this.dataSource.data.splice(index, 1);
          this.dataSource.paginator = this.paginator;
        });
      }
    });
  }

  updateStudent(id: number, firstName: string, lastName: string, gender: string, grade: string, userName: string, password: string) {
    const dialogRef = this.dialog.open(UpdateStudentDialogComponent, {
      data: {firstName, lastName, gender, grade, userName, password },
    });

    dialogRef.afterClosed().subscribe((result) => {
      if (result) {
        const relevantStudent = _.find(this.dataSource.data, {id});
        const tempStudent = _.assign({}, relevantStudent, result);

        this.studentService.update(tempStudent)
          .then((data) => {
            const index = _.findIndex(this.dataSource.data, (student) => student.id === id);
            this.dataSource.data[index] = _.assign({}, relevantStudent, result);
            this.dataSource.paginator = this.paginator;
          });
      }
    });
  }

  _createNewStudent(userData: any): Student {
    const student: Student = new Student();
    student.id = 0;
    student.firstName = userData.firstName;
    student.lastName = userData.lastName;
    student.userName = userData.userName;
    student.class = userData.Class;
    return student;
  }
}
