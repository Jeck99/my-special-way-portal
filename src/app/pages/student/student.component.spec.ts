jest.mock('./services/student.graphql.service');

import { TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { MatHeaderRowDef, MatRowDef, MatHeaderRow, MatDialog } from '@angular/material';
import { StudentComponent } from './student.component';
import { StudentService } from './services/student.graphql.service';

describe('student component', () => {
  beforeEach(async () => {

    class UserDialogMock {
      open = jest.fn().mockImplementation(() => {
        return {
          afterClosed: () => {
            return {
              subscribe: jest.fn(),
            };
          },
        };
      });
    }

    // class UserServiceMock {
    //   getAllUsers = jest.fn().mockImplementation(() => {
    //     return '';
    //   });
    // }

    TestBed.configureTestingModule({
      imports: [

      ],
      declarations: [
        StudentComponent,
        MatHeaderRow,
        MatRowDef,
        MatHeaderRowDef,

      ],
      providers: [
        StudentService,
        { provide: MatDialog, useClass: UserDialogMock },

      ],
      schemas: [NO_ERRORS_SCHEMA],
    });

    // shallow = new Shallow(UserComponent, AppModule);
  });

  it('should render component as described in snapshot', () => {
    const fixture = TestBed.createComponent(StudentComponent);
    expect(fixture).toMatchSnapshot();
  });

  it('should open dialog when calling addNewStudent function', () => {
    const fixture = TestBed.createComponent(StudentComponent);
    fixture.componentInstance.addNewStudent();
    const studentDialogMock = TestBed.get(MatDialog);
    expect(studentDialogMock.open).toHaveBeenCalled();
  });

  it('should open dialog when calling deleteStudent function', () => {
    const fixture = TestBed.createComponent(StudentComponent);
    fixture.componentInstance.deleteStudent(123, 'sad', 'asd', 'אשוח');
    const studentDialogMock = TestBed.get(MatDialog);
    expect(studentDialogMock.open).toHaveBeenCalled();
  });

  it('should open dialog when calling updateeStudent function', () => {
    const fixture = TestBed.createComponent(StudentComponent);
    fixture.componentInstance.updateStudent(123, 'sad', 'asd', 'asd', 'asd', 'asd', 'asd');
    const studentDialogMock = TestBed.get(MatDialog);
    expect(studentDialogMock.open).toHaveBeenCalled();
  });

});
