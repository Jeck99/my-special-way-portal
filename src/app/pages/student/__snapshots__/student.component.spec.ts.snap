// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dashboard component should render component as described in snapshot 1`] = `
<app-student
  dataSource={[Function MatTableDataSource]}
  dialog={[Function StudentDialogMock]}
  displayedColumns={[Function Array]}
  paginator="undefined"
  resultsLength="0"
  sort="undefined"
  studentService={[Function StudentService]}
  table={[Function ElementRef]}
>
  <div
    class="main-content"
  >
    
    
    <div
      class="container-fluid"
    >
      
      
      <div
        class="row"
      >
        
        
        <div
          class="example-header"
        >
          
          
          <mat-form-field
            color="accent"
            floatplaceholder="never"
          >
            
            
            <input
              matinput=""
              placeholder="Filter issues"
            />
            
          
          </mat-form-field>
          
        
        </div>
        
        
        <div
          class="example-header"
        >
          
          
          <button
            color="primary"
            mat-raised-button=""
          >
            הוספת תלמיד חדש
          </button>
          
        
        </div>
        
      
      </div>
      
      
      <div
        class="row"
      >
        
        
        <div
          class="col-lg-12 col-md-12 col-sm-12"
        >
          
          
          <mat-table
            class="example-table"
            matsort=""
            matsortactive="firstName"
            matsortdirection="asc"
            matsortdisableclear=""
          >
            
            
            
            
            
            
              
            
            
              
            
            
            
            
  
            
            
            
            
            
              
            
            
              
            <td>
              
                
              
              
              
            </td>
            
            
            
  
            
            
            
            
            
              
            
            
              
            
            
            
            
  
            
            
            
            
            
              
            
            
              
            
            
            
            
  
            
            
            
            
            
              
            
            
              
            
            
            
            
  
            
            
            
            
            --&gt;
              
            
            
              
            
            
            
            
  
            
            
            
            
            
            
          
          </mat-table>
          
  
          
          <mat-paginator>
            
          
          </mat-paginator>
          
        
        </div>
        
      
      </div>
      
    
    </div>
    
  
  </div>
  
</app-student>
`;
